### PR TITLE
NMS-12316: Run script policies at the end of NodeScan.

### PR DESCRIPTION
**Pull request by cgorantla**
Run ScriptPolicy in transaction at the end of Scan phase before sending scan completed event.
Scripts should use `transaction` variable to see if they need to be run in transaction or not.
Add groovy script for choosing SNMP primary interface by matching ifNames.

### External References

* JIRA (Issue Tracker): https://issues.opennms.org/browse/NMS-12316
* Bamboo (Continuous Integration): https://bamboo.opennms.org/

### CODE REVIEW

**Comment by christianpape** [comment]
I created a requisition and added a node with two interfaces and another one in the 192.168.100.0/24 network. Then I just used the small example groovy-script in etc/script-policies which sets the 192.168.100.x address to primary and all other to secondary. Furthermore the script sets some asset fields and adds a custom monitoring location. But all of this doesn’t seem to work anymore. The logs show that the stuff were called but after creation of the node neither the primary interface was altered nor the assets/location was modified.

**Comment by cgorantla** [comment]
> I created a requisition and added a node with two interfaces and another one in the 192.168.100.0/24 network. Then I just used the small example groovy-script in etc/script-policies which sets the 192.168.100.x address to primary and all other to secondary. Furthermore the script sets some asset fields and adds a custom monitoring location. But all of this doesn’t seem to work anymore. The logs show that the stuff were called but after creation of the node neither the primary interface was altered nor the assets/location was modified.

Thanks for checking this @christianpape . I m making some more changes to handle the case where we run script in multiple places. That should solve the issue you are seeing.

**Comment by cgorantla** [comment]
@christianpape  Can you take a look again.

**Comment by christianpape** [comment]
> @christianpape Can you take a look again.

Sure. I'm building it...

**Comment by christianpape** [comment]
Hi, I created a requisition with two nodes:
* node1: 192.168.31.100 (P), 192.168.100.100 (N), 10.174.24.100 (N)
* node2: 127.0.0.1 (P)
...and the following script:
`import org.opennms.netmgt.model.OnmsIpInterface;
import org.opennms.netmgt.model.monitoringLocations.OnmsMonitoringLocation;
import org.opennms.netmgt.model.PrimaryType;

// Example script for modifying a node's primary interface, assets and the monitoring location.

for(OnmsIpInterface iface : node.getIpInterfaces()) {

    // 192.168.100.0/24 interface should be primary while the remaining should be set to SECONDARY
    if (iface.getIpAddressAsString().matches("^192\\.168\\.100\\..*")) {
        LOG.warn("My-Script: " + iface.getIpAddressAsString() + " set to PRIMARY")
        iface.setIsSnmpPrimary(PrimaryType.PRIMARY)
    } else {
        LOG.warn("My-Script: " + iface.getIpAddressAsString() + " set to SECONDARY")
        iface.setIsSnmpPrimary(PrimaryType.SECONDARY)
    }
    // WORKS!
}

// The location for this node will be set to a custom value.

// setting a custom location
LOG.warn("My-Script: Setting custom location 'Minneapolis'")
node.setLocation(new OnmsMonitoringLocation("Minneapolis", "")); 
// DOES NOT WORK!

LOG.warn("My-Script: Setting operatingSystem to 'macOS'")
node.getAssetRecord().setOperatingSystem("macOS");
// DOES NOT WORK!

return node;
`

The primary interface is set correctly but neither the location nor the assets are set.

**Comment by cgorantla** [comment]
> Hi, I created a requisition with two nodes:
> 
> * node1: 192.168.31.100 (P), 192.168.100.100 (N), 10.174.24.100 (N)
> * node2: 127.0.0.1 (P)
>   ...and the following script:
> 
> ```
> import org.opennms.netmgt.model.OnmsIpInterface;
> import org.opennms.netmgt.model.monitoringLocations.OnmsMonitoringLocation;
> import org.opennms.netmgt.model.PrimaryType;
> 
> // Example script for modifying a node's primary interface, assets and the monitoring location.
> for(OnmsIpInterface iface : node.getIpInterfaces()) {
>     // 192.168.100.0/24 interface should be primary while the remaining should be set to SECONDARY
>     if (iface.getIpAddressAsString().matches("^192\\.168\\.100\\..*")) {
>         LOG.warn("My-Script: " + iface.getIpAddressAsString() + " set to PRIMARY")
>         iface.setIsSnmpPrimary(PrimaryType.PRIMARY)
>     } else {
>         LOG.warn("My-Script: " + iface.getIpAddressAsString() + " set to SECONDARY")
>         iface.setIsSnmpPrimary(PrimaryType.SECONDARY)
>     }
>     // WORKS!
> }
> 
> // setting a custom location
> LOG.warn("My-Script: Setting custom location 'Minneapolis'")
> node.setLocation(new OnmsMonitoringLocation("Minneapolis", "")); 
> // DOES NOT WORK!
> 
> // setting an asset field
> LOG.warn("My-Script: Setting operatingSystem to 'macOS'")
> node.getAssetRecord().setOperatingSystem("macOS");
> // DOES NOT WORK!
> 
> return node;
> ```
> 
> The primary interface is set correctly but neither the location nor the assets are set.

Looks like you need an entry in `monitoringlocations` for setting the location.  Since that is throwing exception assets were not saved.

**Comment by christianpape** [comment]
> > Hi, I created a requisition with two nodes:
> > 
> > * node1: 192.168.31.100 (P), 192.168.100.100 (N), 10.174.24.100 (N)
> > * node2: 127.0.0.1 (P)
> >   ...and the following script:
> > 
> > ```
> > import org.opennms.netmgt.model.OnmsIpInterface;
> > import org.opennms.netmgt.model.monitoringLocations.OnmsMonitoringLocation;
> > import org.opennms.netmgt.model.PrimaryType;
> > 
> > // Example script for modifying a node's primary interface, assets and the monitoring location.
> > for(OnmsIpInterface iface : node.getIpInterfaces()) {
> >     // 192.168.100.0/24 interface should be primary while the remaining should be set to SECONDARY
> >     if (iface.getIpAddressAsString().matches("^192\\.168\\.100\\..*")) {
> >         LOG.warn("My-Script: " + iface.getIpAddressAsString() + " set to PRIMARY")
> >         iface.setIsSnmpPrimary(PrimaryType.PRIMARY)
> >     } else {
> >         LOG.warn("My-Script: " + iface.getIpAddressAsString() + " set to SECONDARY")
> >         iface.setIsSnmpPrimary(PrimaryType.SECONDARY)
> >     }
> >     // WORKS!
> > }
> > 
> > // setting a custom location
> > LOG.warn("My-Script: Setting custom location 'Minneapolis'")
> > node.setLocation(new OnmsMonitoringLocation("Minneapolis", "")); 
> > // DOES NOT WORK!
> > 
> > // setting an asset field
> > LOG.warn("My-Script: Setting operatingSystem to 'macOS'")
> > node.getAssetRecord().setOperatingSystem("macOS");
> > // DOES NOT WORK!
> > 
> > return node;
> > ```
> > 
> > 
> > The primary interface is set correctly but neither the location nor the assets are set.
> 
> Looks like you need an entry in `monitoringlocations` for setting the location. Since that is throwing exception assets were not saved.

Yes, you are right. Works for an existing location!

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/NodeScan.java [hunk 1]
```diff
@@ -458,7 +466,7 @@ public void processIPAddressRow(final IPAddressRow row) {
                         final List<IpInterfacePolicy> policies = getProvisionService().getIpInterfacePoliciesForForeignSource(getForeignSource() == null ? "default" : getForeignSource());
                         for(final IpInterfacePolicy policy : policies) {
                             if (iface != null) {
-                                iface = policy.apply(iface);
+                                iface = policy.apply(iface, new HashMap<>());
```
Comment (j-white) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/NodeScan.java]: Use `Collections.emptyMap()` instead. (For other references too.)

File: opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/policies/ScriptPolicy.java [hunk 2]
```diff
@@ -245,4 +285,12 @@ public String getForeignSource() {
     public void setForeignSource(String foreignSource) {
         putCriteria("foreignSource", foreignSource);
     }
+
+    public void setNodeDao(NodeDao nodeDao) {
+        this.m_nodeDao = nodeDao;
```
Comment (j-white) [inline opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/policies/ScriptPolicy.java]: `this.m_`

File: opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/policies/ScriptPolicy.java [hunk 3]
```diff
@@ -114,25 +124,55 @@ private CompiledScript compileScript(final String script) throws IOException, Sc
...
+                }
             }
-        } catch (Exception ex) {
-            LOG.warn("Error applying ScriptPolicy.", ex);
+        } catch (ScriptException e) {
+            LOG.warn("Error while compiling script.", e);
+        } catch (IOException e) {
+            LOG.warn("Error while opening script file {}.", m_script, e);
+        } catch (Exception e) {
+            LOG.warn("Unkown Error while applying script.", e);
```
Comment (j-white) [inline opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/policies/ScriptPolicy.java]:
```suggestion
            LOG.warn("Unknown error while applying script.", e);
```

File: opennms-base-assembly/src/main/filtered/etc/examples/script-policies/example.groovy [hunk 4]
```diff
@@ -26,29 +26,103 @@
...
+        .orElse(null);
+
+OnmsIpInterface ipInterface = null;
+
+if (matchedIfName == null) {
+    // Find lowest IP Address as there is no match in if names.
+    List<byte[]> ipAddressesInBytes = interfacesWithSNMP.stream().map({ onmsIpInterface -> onmsIpInterface.getIpAddress().getAddress() }).collect(Collectors.toList());
+    byte[] lowestNumberedIp = Collections.min(ipAddressesInBytes, new ByteArrayComparator());
+    ipInterface = interfacesWithSNMP.stream()
+            .filter({ onmsIpInterface -> Arrays.equals(onmsIpInterface.getIpAddress().getAddress(), lowestNumberedIp) })
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/etc/examples/script-policies/example.groovy]: We already ran through the list above, and now we're going through it again to find the match. Can we avoid visiting each IP interface twice?

File: opennms-base-assembly/src/main/filtered/etc/examples/script-policies/example.groovy [hunk 5]
```diff
@@ -26,29 +26,103 @@
...
+            .findFirst().orElse(null);
+    if (ipInterface != null) {
+        LOG.debug("customscript : lowest numbered ipAddress = {}", ipInterface.getIpAddress());
+    }
+
+} else {
+    LOG.debug("customscript : Found a match for ifName = {}", matchedIfName);
+
+    OnmsSnmpInterface result = snmpInterfaces.stream()
+            .filter({ snmpInterface -> snmpInterface.getIfName().equals(matchedIfName) })
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/etc/examples/script-policies/example.groovy]: Similarly here for SNMP interfaces. Nodes can have hundreds of IP interfaces, and thousands of SNMP interfaces.

**Pull request merged**